flow/Deque.h

```diff
@@ -41,21 +41,25 @@ class Deque {
 	Deque() : arr(0), begin(0), end(0), mask(-1) {}
 
 	// TODO: iterator construction, other constructors
-	Deque(Deque const& r) : arr(0), begin(0), end(r.size()), mask(r.mask) {
+	Deque(Deque const& r) : arr(nullptr), begin(0), end(r.size()), mask(r.mask) {
 		if (r.capacity() > 0) {
 			arr = (T*)aligned_alloc(std::max(__alignof(T), sizeof(void*)), capacity() * sizeof(T));
 			ASSERT(arr != nullptr);
 		}
 		ASSERT(capacity() >= end || end == 0);
-		for (uint32_t i=0; i<end; i++)
-			new (&arr[i]) T(r[i]);
-		// FIXME: Specialization for POD types using memcpy?
+		if (r.end < r.capacity()) {
+			std::copy(r.arr + r.begin, r.arr + r.begin + r.size(), arr);
+		} else {
+			auto partOneSize = r.capacity() - r.begin;
+			std::copy(r.arr + r.begin, r.arr + r.begin + partOneSize, arr);
+			std::copy(r.arr, r.arr + (r.end & r.mask), arr + partOneSize);
```

Comment:
would you mind adding a comment explaining that `end` is always larger than `begin` and the "mask" is used to decide the queue's end;
The comment will do the future ourselves a favor to avoid the misunderstanding of how the `end` works. :)